### DISCUSSION TITLE
Should WezTerm warn about "Private Use Area" glyphs?

### DISCUSSION THREAD

**Discussion by karypid** [wezterm/wezterm#4377] [General]
Hello,

I just installed WezTerm on my EndeavourOS (Arch-based) Linux system.

First thing that came up when I tried to use `yay` to update my system is that the terminal warned me that it was missing the ability to display some character with:

![Screenshot from 2023-10-01 13-39-17](https://github.com/wez/wezterm/assets/1221101/ef3795e7-1ca5-4927-9d20-9e1fc12c395e)

After reading those docs, I added some more monospace fonts to the configuration, thinking that the "default bundled fonts" are missing some glyph. Following those instructions, I added Liberation and Nimbus PS:

```
$ wezterm ls-fonts
Primary font:
wezterm.font_with_fallback({
  -- /usr/share/fonts/liberation/LiberationMono-Regular.ttf, FontConfig
  "Liberation Mono",

  -- /usr/share/fonts/gsfonts/NimbusMonoPS-Regular.otf, FontConfig
  "Nimbus Mono PS",

  -- /usr/share/fonts/TTF/JetBrainsMono-Regular.ttf, FontConfig
  "JetBrains Mono",

})
```

The problem is that even with these 2 extra fonts, I still get the warning. So I looked up the character and [found this information in codepoints](https://codepoints.net/U+F023) which explains that it is a "private use area" character which means:

> In Unicode, a **Private Use Area** (**PUA**) is a range of code points that, by definition, will not be assigned characters by the Unicode Consortium.

> They are intentionally left undefined so that third parties may define their own characters without conflicting with Unicode Consortium assignments.

This is what WizTerm displays, I am assuming that the character is the red-circled one?

![Screenshot from 2023-10-01 14-08-13](https://github.com/wez/wezterm/assets/1221101/447fb6c8-1868-43ab-a950-bc330c1e5d7b)

But, I'm confused now: how do I fix this?

At first I thought it was missing from the fonts used, so I added fonts.

Now, it seems to be the case that `yay` is using a "private character" that is non standard when prompting for a password?

**Comment by wez** [comment]
Font aggregators such as Nerd Fonts make heavy use of private use codepoints for their glyphs.  Most users are not aware that this is even a thing when it comes to text. The goal of the warning is to advise you that you are not able to see the content you are trying to display.

You can use https://wezfurlong.org/wezterm/config/lua/config/warn_about_missing_glyphs.html to disable the warning if you wish.

**Reply by karypid** [reply]
So, help me understand this (will make this discussion useful for others). I understand that:

- there are some unicode ranges(PUA)  left to "anyone" to define 
- some web technologies (devicons, font awesome) use those ranges to add icons as font characters

Having said that, I would assume that a web page using that font, would also have you download the web-font: this makes sure that when they display unicode character X in the PUA range, you see what they intended: they have specified the respective glyph in their own web font, right?

Now in this case, the `yay` tool is trying to display a character in the PUA range, right? That character could be mapped to anything depending on which font is actually used? So what is the "proper" way to display this character? The only way to ensure you see the correct thing, would be to use a font that the `yay` developers use, where `\u{f023}` contains the glyph they intended?

When I run this in other terminals (foot / gnome-terminal / terminology) I see nothing special on the screen (just the text in the screenshot). I really appreciate how wezterm is pointing this out.

Now, it seems that in my case I do not have the nerd fonts on my system at all. Which ones do you normally bundle with wezterm? Here are the ones that seem to be monospace:

```
$ yay -Ss nerd-fonts | grep -v "^ " | grep mono
aur/otf-nerd-fonts-monacob-mono 3.0.2-1 (+1 0.01)
aur/otf-operator-mono-nerd 1.200-3 (+1 0.00) (Orphaned)
aur/nerd-fonts-noto-sans-mono-extended 3.0.2-2 (+2 0.87)
aur/nerd-fonts-jetbrains-mono-160 2.1.0-1 (+1 0.12) (Out-of-date: 2022-08-26)
aur/nerd-fonts-sf-mono-ligatures 2.2.2-1 (+0 0.00) (Out-of-date: 2023-01-18)
aur/nerd-fonts-complete-mono-glyphs 2.1.0-1 (+15 0.00) (Orphaned) (Out-of-date: 2022-08-26)
aur/nerd-fonts-dm-mono 1.000r5.57fadab_2.2.0_RC-1 (+0 0.00)
aur/nerd-fonts-sf-mono 3.0.2-1 (+9 0.94)
extra/ttf-victor-mono-nerd 3.0.2-1 (6.9 MiB 120.7 MiB) [nerd-fonts]
extra/otf-firamono-nerd 3.0.2-1 (4.0 MiB 43.5 MiB) [nerd-fonts]
extra/ttf-ibmplex-mono-nerd 3.0.2-1 (3.5 MiB 87.7 MiB) [nerd-fonts]
extra/ttf-space-mono-nerd 3.0.2-1 (1.8 MiB 21.5 MiB) [nerd-fonts]
extra/ttf-liberation-mono-nerd 3.0.2-1 (4.2 MiB 59.7 MiB) [nerd-fonts]
extra/ttf-daddytime-mono-nerd 3.0.2-1 (1.6 MiB 5.4 MiB) [nerd-fonts]
extra/ttf-sharetech-mono-nerd 3.0.2-1 (1.5 MiB 5.2 MiB) [nerd-fonts]
extra/ttf-monoid-nerd 3.0.2-1 (1.9 MiB 22.0 MiB) [nerd-fonts]
extra/ttf-jetbrains-mono-nerd 3.0.2-1 (6.1 MiB 184.1 MiB) [nerd-fonts]
extra/ttf-mononoki-nerd 3.0.2-1 (1.8 MiB 21.8 MiB) [nerd-fonts]
extra/ttf-nerd-fonts-symbols-mono 3.0.2-2 (987.0 KiB 1.8 MiB) [nerd-fonts]
extra/ttf-monofur-nerd 3.0.2-1 (2.0 MiB 17.5 MiB) [nerd-fonts]
extra/ttf-bitstream-vera-mono-nerd 3.0.2-1 (1.9 MiB 22.0 MiB) [nerd-fonts]
extra/ttf-roboto-mono-nerd 3.0.2-1 (3.8 MiB 67.6 MiB) [nerd-fonts]
extra/ttf-ubuntu-mono-nerd 3.0.2-1 (2.0 MiB 22.3 MiB) [nerd-fonts]
```

I'm hoping that if I add one of these I will see "something", not necessarily the correct thing though?

**Reply by karypid** [reply]
Update: I installed `extra/ttf-liberation-mono-nerd` and now I see a lock icon:

![image](https://github.com/wez/wezterm/assets/1221101/9f11d665-b035-4391-b319-669bffc58e0e)

This does make sense. I wonder how the font got automatically picked up though. Currently I have:

```
$ cat .config/wezterm/wezterm.lua
-- Pull in the wezterm API
local wezterm = require 'wezterm'

-- add more system fonts
wezterm.font("Liberation Mono", {weight="Regular", stretch="Normal", style="Normal"}) -- /usr/share/fonts/liberation/LiberationMono-Regular.ttf, FontConfig
wezterm.font("Liberation Mono", {weight="Regular", stretch="Normal", style="Italic"}) -- /usr/share/fonts/liberation/LiberationMono-Italic.ttf, FontConfig
wezterm.font("Liberation Mono", {weight="Bold", stretch="Normal", style="Normal"}) -- /usr/share/fonts/liberation/LiberationMono-Bold.ttf, FontConfig
wezterm.font("Liberation Mono", {weight="Bold", stretch="Normal", style="Italic"}) -- /usr/share/fonts/liberation/LiberationMono-BoldItalic.ttf, FontConfig
wezterm.font("Nimbus Mono PS", {weight="Regular", stretch="Normal", style="Normal"}) -- /usr/share/fonts/gsfonts/NimbusMonoPS-Regular.otf, FontConfig
wezterm.font("Nimbus Mono PS", {weight="Regular", stretch="Normal", style="Italic"}) -- /usr/share/fonts/gsfonts/NimbusMonoPS-Italic.otf, FontConfig
wezterm.font("Nimbus Mono PS", {weight="Bold", stretch="Normal", style="Normal"}) -- /usr/share/fonts/gsfonts/NimbusMonoPS-Bold.otf, FontConfig
wezterm.font("Nimbus Mono PS", {weight="Bold", stretch="Normal", style="Italic"}) -- /usr/share/fonts/gsfonts/NimbusMonoPS-BoldItalic.otf, FontConfig

-- This table will hold the configuration.
local config = {}

... (rest of config)
```

Do I not need to explicitly add the font?

```
$ wezterm ls-fonts | grep -i nerd
(zero output)

$ wezterm ls-fonts --list-system | grep -i nerd
wezterm.font("LiterationMono Nerd Font", {weight="Regular", stretch="Normal", style="Normal"}) -- /usr/share/fonts/TTF/LiterationMonoNerdFont-Regular.ttf, FontConfig
wezterm.font("LiterationMono Nerd Font", {weight="Regular", stretch="Normal", style="Italic"}) -- /usr/share/fonts/TTF/LiterationMonoNerdFont-Italic.ttf, FontConfig
wezterm.font("LiterationMono Nerd Font", {weight="Bold", stretch="Normal", style="Normal"}) -- /usr/share/fonts/TTF/LiterationMonoNerdFont-Bold.ttf, FontConfig
wezterm.font("LiterationMono Nerd Font", {weight="Bold", stretch="Normal", style="Italic"}) -- /usr/share/fonts/TTF/LiterationMonoNerdFont-BoldItalic.ttf, FontConfig
... (more fonts
```

**Reply by wez** [reply]
font technology doesn't know or care about PUA; it just tries to resolve the codepoint from a font.

wezterm includes `Symbols Nerd Font Mono` in its default fallback.
If you installed from the arch package, the person who maintains the arch packaging has decided not to include that font, so the behavior may be different there; I would expect to see this at the end of your `ls-fonts` output:

```
  -- <built-in>, BuiltIn
  "Symbols Nerd Font Mono",
```

You can try adding in `Symbols Nerd Font mono` to your `wezterm.font_with_fallback` in your config.

Separately, wezterm will use font-config to find fonts based on codepoint when none of your listed fonts match.

It will perform the equivalent of this command line query to find a match:

```console
$ fc-list ':charset=f023'
/usr/share/fonts/fontawesome/FontAwesome.otf: FontAwesome:style=Regular
/home/wez/.local/share/fonts/IosevkaNerdFont-Bold.ttf: Iosevka Nerd Font,Iosevka NF:style=Bold
/home/wez/.fonts/Font Awesome 5 Free-Solid-900.otf: Font Awesome 5 Free,Font Awesome 5 Free Solid:style=Solid
/usr/share/fonts/fontawesome/fontawesome-webfont.ttf: FontAwesome:style=Regular
/home/wez/.local/share/fonts/IosevkaTermNerdFont-Bold.ttf: IosevkaTerm Nerd Font,IosevkaTerm NF:style=Bold
```

Another complication is that Nerd Fonts recently made incompatible changes to their PUA area in the move from 2.0 to 3.0.

If you have the 2.0 font and yay (or whatever) is assuming 3.0, then you may not find the right glyph, and vice-versa.

**Reply by karypid** [reply]
You are absolutely right. I've installed the fonts mentioned as bundled in the docs and now I see:

```
Primary font:
wezterm.font_with_fallback({
  -- /usr/share/fonts/TTF/JetBrainsMono-Regular.ttf, FontConfig
  "JetBrains Mono",

  -- /usr/share/fonts/TTF/SymbolsNerdFontMono-Regular.ttf, FontConfig
  "Symbols Nerd Font Mono",

  -- /usr/share/fonts/noto/NotoColorEmoji.ttf, FontConfig
  -- Assumed to have Emoji Presentation
  -- Pixel sizes: [128]
  "Noto Color Emoji",

})
```

I also don't get the warning anymore. I've got a post about this in the EndeavourOS forums. I will update it referencing the treasure trove of information here.

The warning was really useful! Glad I didn't just ignore it.

Thank you!
